### PR TITLE
Bug/producer consumer adminclient validity handling

### PR DESCRIPTION
**Pull request by PrasanthV454**
- Added isClosed new field to Producer struct which acts as flag to know if Producer is open or closed.

- Added new function verifyProducer() function which checks if producer is valid or not by doing null checks and closed checks to avoid seg faults for C-GO or go-panics.

### CODE REVIEW

File: kafka/adminapi.go [hunk 1]
```diff
@@ -622,6 +629,10 @@ func (a *AdminClient) waitResult(ctx context.Context, cQueue *C.rd_kafka_queue_t
 
 // cToTopicResults converts a C topic_result_t array to Go TopicResult list.
 func (a *AdminClient) cToTopicResults(cTopicRes **C.rd_kafka_topic_result_t, cCnt C.size_t) (result []TopicResult, err error) {
+	err = a.verifyClient()
```
Comment (milindl) [inline kafka/adminapi.go:632]: Not required to check this in non-exposed/non-public functions (here and elsewhere)

Comment (milindl) [reply kafka/adminapi.go:632]: All the conversion c->go, go->c methods etc all don't need this

File: kafka/adminapi_test.go [hunk 2]
```diff
@@ -30,26 +30,30 @@ func TestAdminAPIWithDefaultValue(t *testing.T) {
...
 	}
 
-	expDuration, err := time.ParseDuration("30s")
+	adminClient, err := NewAdminClient(&conf)
 	if err != nil {
-		t.Fatalf("Failed to Parse Duration: %s", err)
+		t.Fatalf("Failed to create AdminClient %v", err)
 	}
+	testAdminAPIWithDefaultValue(t, adminClient, nil)
+	testAdminAPIWithDefaultValue(t, adminClient, getOperationNotAllowedErrorForClosedClient())
```
Comment (milindl) [inline kafka/adminapi_test.go:43]: Add comment here that `adminClient` is closed by the time we call the second one, because otherwise it requires one to read the whole other function.

File: kafka/adminapi_test.go [hunk 3]
```diff
@@ -58,22 +62,27 @@ func TestAdminAPIWithDefaultValue(t *testing.T) {
...
+	if !a.isClosed {
+		t.Logf("Succeed to create topic %v\n", res)
 	}
-	t.Logf("Succeed to create topic %v\n", res)
 
 	ctx, cancel = context.WithTimeout(context.Background(), expDuration)
 	defer cancel()
-	res, err = adminClient.DeleteTopics(ctx, []string{topic})
+	res, err = a.DeleteTopics(ctx, []string{topic})
 	if err != nil {
```
Comment (milindl) [inline kafka/adminapi_test.go:79]: `err != errCheck` here as well?

**Review by milindl — COMMENTED** [review]
Looks much cleaner than last time

File: kafka/adminapi.go [hunk 4]
```diff
@@ -703,6 +718,11 @@ func (a *AdminClient) ClusterID(ctx context.Context) (clusterID string, err erro
 //
 // Requires broker version >= 0.10.0.
 func (a *AdminClient) ControllerID(ctx context.Context) (controllerID int32, err error) {
+	err = a.verifyClient()
+	if err != nil {
+		return 0, err
```
Comment (edenhill) [inline kafka/adminapi.go:723]: -1
I believe 0 is a valid broker id

File: kafka/adminapi.go [hunk 5]
```diff
@@ -1555,15 +1641,21 @@ func (a *AdminClient) SetSaslCredentials(username, password string) error {
 
 // Close an AdminClient instance.
 func (a *AdminClient) Close() {
+	if a.isClosed {
+		return
+	}
+
```
Comment (edenhill) [inline kafka/adminapi.go:1647]: To avoid a race condition this should be a cmp and exchange, or perhaps an increment (if they are atomic in go).
Two different threads calling Close() at the same time now risks the below code to execute in parallel and thus crash.

File: kafka/adminapi_test.go [hunk 6]
```diff
@@ -58,22 +65,32 @@ func TestAdminAPIWithDefaultValue(t *testing.T) {
...
-		adminClient.Close()
-		t.Fatalf("Failed to create topics %v\n", err)
+	if err != errCheck {
+		t.Errorf("CreateTopics() failed to return expected result")
+		if !a.isClosed {
+			a.Close()
+			t.Fatalf("Failed to create topics %v\n", err)
+		}
+	}
+	if !a.isClosed {
```
Comment (edenhill) [inline kafka/adminapi_test.go:75]: Why are you checking isClosed here and below?

Also, these tests should preferably only use the public APIs and isClosed is not.

File: kafka/consumer.go [hunk 7]
```diff
@@ -308,25 +365,28 @@ func (c *Consumer) StoreMessage(m *Message) (storedOffsets []TopicPartition, err
...
+	err := c.verifyClient()
+	if err != nil {
+		return err
+	}
 	rkt := c.handle.getRkt(*partition.Topic)
 	cErr := C.rd_kafka_seek(rkt,
 		C.int32_t(partition.Partition),
 		C.int64_t(partition.Offset),
-		C.int(-1))
+		C.int(timeoutMs))
```
Comment (edenhill) [inline kafka/consumer.go:389]: these changes are not related to this PR, should not be here

File: kafka/consumer.go [hunk 8]
```diff
@@ -442,6 +506,8 @@ func (c *Consumer) Close() (err error) {
 
 	C.rd_kafka_destroy(c.handle.rk)
 
+	c.isClosed = true
```
Comment (edenhill) [inline kafka/consumer.go:509]: same thing here, race condition

File: kafka/consumer.go [hunk 9]
```diff
@@ -575,11 +642,19 @@ func consumerReader(c *Consumer, termChan chan bool) {
...
+		return nil, err
+	}
 	return getMetadata(c, topic, allTopics, timeoutMs)
 }
 
 // QueryWatermarkOffsets queries the broker for the low and high offsets for the given topic and partition.
 func (c *Consumer) QueryWatermarkOffsets(topic string, partition int32, timeoutMs int) (low, high int64, err error) {
+	err = c.verifyClient()
+	if err != nil {
+		return 0, 0, err
```
Comment (edenhill) [inline kafka/consumer.go:656]: use -1 for invalid value, 0 is a valid offset.

File: kafka/consumer.go [hunk 10]
```diff
@@ -588,6 +663,10 @@ func (c *Consumer) QueryWatermarkOffsets(topic string, partition int32, timeoutM
 // The low offset is populated every statistics.interval.ms if that value is set.
 // OffsetInvalid will be returned if there is no cached offset for either value.
 func (c *Consumer) GetWatermarkOffsets(topic string, partition int32) (low, high int64, err error) {
+	err = c.verifyClient()
+	if err != nil {
+		return 0, 0, err
```
Comment (edenhill) [inline kafka/consumer.go:668]: use -1

File: kafka/consumer_test.go [hunk 11]
```diff
@@ -189,45 +194,50 @@ func TestConsumerAPIs(t *testing.T) {
...
 
 	err = c.Close()
-	if err != nil {
+	if err != errCheck {
 		t.Errorf("Close failed: %s", err)
 	}
 
 	// Tests the SetSaslCredentials call to ensure that the API does not crash.
-	c.SetSaslCredentials("username", "password")
+	if !c.isClosed {
```
Comment (edenhill) [inline kafka/consumer_test.go:238]: Why isn't this tested when closed?

File: kafka/error.go [hunk 12]
```diff
@@ -163,3 +163,7 @@ func getFatalError(H Handle) error {
 func testFatalError(H Handle, code ErrorCode, str string) ErrorCode {
 	return ErrorCode(C.rd_kafka_test_fatal_error(H.gethandle().rk, C.rd_kafka_resp_err_t(code), C.CString(str)))
 }
+
+func getOperationNotAllowedErrorForClosedClient() error {
+	return newErrorFromString(ErrState, "Operation not allowed on closed client.")
```
Comment (edenhill) [inline kafka/error.go:168]: error strings don't have ending "."

File: kafka/producer.go [hunk 13]
```diff
@@ -368,6 +382,8 @@ func (p *Producer) Close() {
 	p.handle.cleanup()
 
 	C.rd_kafka_destroy(p.handle.rk)
+
+	p.isClosed = true
```
Comment (edenhill) [inline kafka/producer.go:386]: this Close() should also check that isClosed is not already done, and fix that race condition

File: kafka/producer.go [hunk 14]
```diff
@@ -643,12 +664,20 @@ func poller(p *Producer, termChan chan bool) {
...
+	}
 	return getMetadata(p, topic, allTopics, timeoutMs)
 }
 
 // QueryWatermarkOffsets returns the broker's low and high offsets for the given topic
 // and partition.
 func (p *Producer) QueryWatermarkOffsets(topic string, partition int32, timeoutMs int) (low, high int64, err error) {
+	err = p.verifyClient()
+	if err != nil {
+		return 0, 0, err
```
Comment (edenhill) [inline kafka/producer.go:679]: -1

File: kafka/producer_test.go [hunk 15]
```diff
@@ -180,10 +180,78 @@ func TestProducerAPIs(t *testing.T) {
 		t.Errorf("OffsetsForTimes() failed but returned non-nil Offsets: %s\n", offsets)
 	}
 
+	p.Close()
```
Comment (edenhill) [inline kafka/producer_test.go:183]: The SetSaslCredentials call must be done before Close

**Review by edenhill — CHANGES_REQUESTED** [review]

File: kafka/adminapi.go (hunk 5 above)
Comment (PrasanthV454) [reply kafka/adminapi.go:1647]: got it. will use atomic.CompareAndSwapUint32()

File: kafka/adminapi_test.go (hunk 6 above)
Comment (PrasanthV454) [reply kafka/adminapi_test.go:75]: We are performing same tests on closed admin client and open admin client. In case of closed client we don't need to close the client but for other case we were closing the client and producing fatal error. So to differentiate a.isClosed helps.

File: kafka/consumer.go (hunk 7 above)
Comment (PrasanthV454) [reply kafka/consumer.go:389]: Reverted.

File: kafka/adminapi.go [hunk 16]
```diff
@@ -638,6 +645,10 @@ func (a *AdminClient) cToTopicResults(cTopicRes **C.rd_kafka_topic_result_t, cCn
 
 // cConfigResourceToResult converts a C ConfigResource result array to Go ConfigResourceResult
 func (a *AdminClient) cConfigResourceToResult(cRes **C.rd_kafka_ConfigResource_t, cCnt C.size_t) (result []ConfigResourceResult, err error) {
+	err = a.verifyClient()
```
Comment (milindl) [inline kafka/adminapi.go:648]: Remove from this function, it's private to our usage

File: kafka/adminapi.go [hunk 17]
```diff
@@ -1190,11 +1244,20 @@ func (a *AdminClient) SetOAuthBearerToken(oauthBearerToken OAuthBearerToken) err
...
+	err := a.verifyClient()
+	if err != nil {
+		return err
+	}
 	return a.handle.setOAuthBearerTokenFailure(errstr)
 }
 
 // aclBindingToC converts a Go ACLBinding struct to a C rd_kafka_AclBinding_t
 func (a *AdminClient) aclBindingToC(aclBinding *ACLBinding, cErrstr *C.char, cErrstrSize C.size_t) (result *C.rd_kafka_AclBinding_t, err error) {
+	err = a.verifyClient()
```
Comment (milindl) [inline kafka/adminapi.go:1256]: Remove from this function, it's private to our usage

File: kafka/adminapi.go [hunk 18]
```diff
@@ -1230,6 +1293,11 @@ func (a *AdminClient) aclBindingToC(aclBinding *ACLBinding, cErrstr *C.char, cEr
 
 // aclBindingFilterToC converts a Go ACLBindingFilter struct to a C rd_kafka_AclBindingFilter_t
 func (a *AdminClient) aclBindingFilterToC(aclBindingFilter *ACLBindingFilter, cErrstr *C.char, cErrstrSize C.size_t) (result *C.rd_kafka_AclBindingFilter_t, err error) {
+	err = a.verifyClient()
```
Comment (milindl) [inline kafka/adminapi.go:1296]: Remove from this function, it's private to our usage

File: kafka/adminapi.go [hunk 19]
```diff
@@ -1291,6 +1359,11 @@ func (a *AdminClient) cToACLBindings(cACLBindings **C.rd_kafka_AclBinding_t, acl
 
 // cToCreateACLResults converts a C acl_result_t array to Go CreateACLResult list.
 func (a *AdminClient) cToCreateACLResults(cCreateAclsRes **C.rd_kafka_acl_result_t, aclCnt C.size_t) (result []CreateACLResult, err error) {
+	err = a.verifyClient()
```
Comment (milindl) [inline kafka/adminapi.go:1362]: Remove from this function, it's private to our usage

File: kafka/adminapi.go [hunk 20]
```diff
@@ -1555,6 +1642,11 @@ func (a *AdminClient) SetSaslCredentials(username, password string) error {
 
 // Close an AdminClient instance.
 func (a *AdminClient) Close() {
+	var newValue uint32 = 1
+	currentValue := atomic.LoadUint32((*uint32)(unsafe.Pointer(&a.isClosed)))
```
Comment (milindl) [inline kafka/adminapi.go:1646]: While this code is logical, I'm not sure about the cast to 4 byte int. Size of boolean in go (at least on my machine) is only 1 byte, and not 4. Since there is no CAS for booleans, maybe we can use uint8 as the type of isClosed, with a comment. 
Other alternatives also welcome on this, can't figure out what is the 100% correct way.

File: kafka/adminapi_test.go [hunk 21]
```diff
@@ -58,22 +65,32 @@ func TestAdminAPIWithDefaultValue(t *testing.T) {
...
 
 	ctx, cancel = context.WithTimeout(context.Background(), expDuration)
 	defer cancel()
-	res, err = adminClient.DeleteTopics(ctx, []string{topic})
-	if err != nil {
-		adminClient.Close()
-		t.Fatalf("Failed to delete topic %v, err: %v", topic, err)
+	res, err = a.DeleteTopics(ctx, []string{topic})
+	if err != errCheck {
+		t.Errorf("DeleteTopics() failed to return expected result")
```
Comment (milindl) [inline kafka/adminapi_test.go:83]: Print both err and errCheck (in this case, and other cases where expected error is being used)

File: kafka/consumer_test.go [hunk 22]
```diff
@@ -45,11 +44,17 @@ func TestConsumerAPIs(t *testing.T) {
...
+	testConsumerAPIs(t, c, nil)
+
+	// testing the API's with closed consumer.
+	testConsumerAPIs(t, c, getOperationNotAllowedErrorForClosedClient())
+}
+
+func testConsumerAPIs(t *testing.T, c *Consumer, errCheck error) {
+	err := c.Subscribe("gotest", nil)
+	if err != errCheck {
 		t.Errorf("Subscribe failed: %s", err)
```
Comment (milindl) [inline kafka/consumer_test.go:58]: ... Failed to return expected result %v : %v\n", errCheck, err

^ something like this here and in other instances

**Review by milindl — CHANGES_REQUESTED** [review]
Almost done, mostly minor things left.

File: kafka/consumer_test.go (hunk 22 above)
Comment (milindl) [reply kafka/consumer_test.go:58]: It's already been done in case of producer tests, you can use that too

File: kafka/adminapi_test.go (hunk 6 above)
Comment (PrasanthV454) [reply kafka/adminapi_test.go:75]: @milindl need a feedback here.

Comment (milindl) [reply kafka/adminapi_test.go:75]: I can't see a way around checking it, if we should only use the public APIs then we can add an `IsClosed() bool` for that.

File: kafka/consumer.go [hunk 23]
```diff
@@ -494,6 +494,10 @@ func (c *Consumer) ReadMessage(timeout time.Duration) (*Message, error) {
 // Close Consumer instance.
 // The object is no longer usable after this call.
 func (c *Consumer) Close() (err error) {
+	if c.IsClosed() {
+		return getOperationNotAllowedErrorForClosedClient()
+	}
+
 	var newValue uint32 = 1
 	currentValue := atomic.LoadUint32(&c.isClosed)
 	if !atomic.CompareAndSwapUint32(&c.isClosed, currentValue, newValue) {
```
Comment (emasab) [inline kafka/consumer.go:503]: You should add `currentValue == 1 || ` here and remove isClosed before because if two threads reach line 501, then one of them reaches line 508 and the second executes line 502 and 503 will execute a compare and swap 1 -> 1 that returns true (the value hasn't changed since when it was read) and continues so it's closing two times

**Review by emasab — CHANGES_REQUESTED** [review]

File: kafka/adminapi.go [hunk 24]
```diff
@@ -103,7 +104,23 @@ import "C"
...
 type AdminClient struct {
 	handle    *handle
-	isDerived bool // Derived from existing client handle
+	isDerived bool   // Derived from existing client handle
+	isClosed  uint32 // to check if Admin Client is closed or not.
+}
+
+// IsClosed returns boolean representing if client is closed or not
+func (a *AdminClient) IsClosed() bool {
+	if a.isClosed == 0 {
```
Comment (emasab) [inline kafka/adminapi.go:113]:
```suggestion
	return atomic.LoadUint32(&a.isClosed) == 1
```

also for other `IsClosed()`

File: kafka/consumer_test.go [hunk 25]
```diff
@@ -176,10 +179,12 @@ func TestConsumerAPIs(t *testing.T) {
 	// OffsetsForTimes
 	offsets, err := c.OffsetsForTimes([]TopicPartition{{Topic: &topic, Offset: 12345}}, 100)
 	t.Logf("OffsetsForTimes() returned Offsets %s and error %s\n", offsets, err)
-	if err == nil {
+	if !c.IsClosed() && err == nil {
 		t.Errorf("OffsetsForTimes() should have failed\n")
+	} else if c.IsClosed() && err != errCheck {
+		t.Errorf("OffsetsForTimes() should have thrown %v but throwed %v", errCheck, err)
```
Comment (emasab) [inline kafka/consumer_test.go:185]: "threw"

File: kafka/consumer.go (hunk 23 above)
Comment (PrasanthV454) [reply kafka/consumer.go:503]: does it still solve our problem? 
because  `currentValue == 1 || !atomic.CompareAndSwapUint32(&c.isClosed, currentValue, newValue)` is still not atomic. Can we instead introduce a mutex to perform this entirely in atomic fashion ?

Comment (emasab) [reply kafka/consumer.go:503]: Why it's not atomic? If it's already 1 it doesn't need to try to set to 1.

To further simplify, you can replace with:

```suggestion
	if !atomic.CompareAndSwapUint32(&c.isClosed, 0, 1) {
```

without anything else, it's the only case where it has to succeed.

Comment (PrasanthV454) [reply kafka/consumer.go:503]: got it.

File: kafka/consumer.go [hunk 26]
```diff
@@ -420,6 +491,10 @@ func (c *Consumer) ReadMessage(timeout time.Duration) (*Message, error) {
 // Close Consumer instance.
 // The object is no longer usable after this call.
 func (c *Consumer) Close() (err error) {
+	currentValue := atomic.LoadUint32(&c.isClosed)
```
Comment (emasab) [inline kafka/consumer.go:494]: This line is not needed, just leave below:

```suggestion
	if !atomic.CompareAndSwapUint32(&c.isClosed, 0, 1) {
		return getOperationNotAllowedErrorForClosedClient()
	}
```

if isClosed is already 1 it'll return false

File: kafka/consumer_test.go (hunk 25 above)
Comment (milindl) [reply kafka/consumer_test.go]: nit: 
thrown -> threw

Comment (milindl) [reply kafka/consumer_test.go]: nit:
throwed->threw, not thrown to threw. Or so I hope :sweat_smile:

File: kafka/adminapi.go [hunk 27]
```diff
@@ -1841,6 +1913,9 @@ func (a *AdminClient) SetSaslCredentials(username, password string) error {
 
 // Close an AdminClient instance.
 func (a *AdminClient) Close() {
+	if !atomic.CompareAndSwapUint32(&a.isClosed, 0, 1) {
+		return
+	}
 	if a.isDerived {
 		// Derived AdminClient needs no cleanup.
 		a.handle = &handle{}
```
Comment (milindl) [inline kafka/adminapi.go:1921]: Due to rebase I guess, some methods now lack the guard 
For example, ListConsumerGroups,  DescribeConsumerGroups and so on

**Review by milindl — APPROVED** [review]
LGTM but please add the missing prelude in the methods that were added after this PR

**Review by emasab — APPROVED** [review]
LGTM

**Comment by nkostoulas** [comment]
hey - I think this is causing some causes when calling `Close` on the Consumer. The Consumer will `Poll` and receive a `RevokedPartitions` event and call the rebalance CB of the application. If the rebalance CB makes a call to `Unassign` this will fail as the consumer is already closed. cc @PrasanthV454 @emasab can you verify?

**Comment by milindl** [comment]
Right, that seems to be the case. I'll file an issue and take a look at this.

**Comment by nkostoulas** [comment]
@milindl is there any reason to call `Unassign` once the Consumer is closed? I figured we could just check in the rebalance CB whether the Consumer has closed (using `IsClosed`) and just ignore the call if that is the case.

**Comment by milindl** [comment]
@nkostoulas in the case of Unassign, it is okay. Especially since Unassign doesn't need to be called explicitly (the library takes care of it if you don't call it), and you can use the check as you mentioned if you do want to call the method. 

However, the issue is bigger, since none of the Commit methods can be called during the closing rebalance, that's problematic for users depending on manual commits. 

Thanks for pointing this out
